processing/src/main/java/io/druid/query/groupby/epinephelinae/StreamingMergeSortedGrouper.java

```diff
@@ -0,0 +1,385 @@
+      final AggregatorFactory[] aggregatorFactories
+  )
+  {
+    this(bufferSupplier, keySerde, columnSelectorFactory, aggregatorFactories, DEFAULT_TIMEOUT_MS);
+  }
+
+  @VisibleForTesting
+  StreamingMergeSortedGrouper(
+      final Supplier<ByteBuffer> bufferSupplier,
+      final KeySerde<KeyType> keySerde,
+      final ColumnSelectorFactory columnSelectorFactory,
+      final AggregatorFactory[] aggregatorFactories,
+      final long timeoutMs
+  )
+  {
+    this.bufferSupplier = bufferSupplier;
+    this.keySerde = keySerde;
+    this.aggregators = new BufferAggregator[aggregatorFactories.length];
+    this.aggregatorOffsets = new int[aggregatorFactories.length];
+
+    this.keySize = keySerde.keySize();
+    int offset = keySize;
+    for (int i = 0; i < aggregatorFactories.length; i++) {
+      aggregators[i] = aggregatorFactories[i].factorizeBuffered(columnSelectorFactory);
+      aggregatorOffsets[i] = offset;
+      offset += aggregatorFactories[i].getMaxIntermediateSize();
+    }
+    this.recordSize = offset;
+    this.timeoutMs = timeoutMs;
+  }
+
+  @Override
+  public void init()
+  {
+    if (!initialized) {
+      buffer = bufferSupplier.get();
+      maxSlotNum = buffer.capacity() / recordSize;
+      Preconditions.checkState(
+          maxSlotNum > 2,
+          "Buffer[%s] should be large enough to store at least three records[%s]",
+          buffer.capacity(),
+          recordSize
+      );
+
+      reset();
+      initialized = true;
+    }
+  }
+
+  @Override
+  public boolean isInitialized()
+  {
+    return initialized;
+  }
+
+  @Override
+  public AggregateResult aggregate(KeyType key, int notUsed)
+  {
+    return aggregate(key);
+  }
+
+  @Override
+  public AggregateResult aggregate(KeyType key)
+  {
+    try {
+      final ByteBuffer keyBuffer = keySerde.toByteBuffer(key);
+
+      if (keyBuffer.remaining() != keySize) {
+        throw new IAE(
+            "keySerde.toByteBuffer(key).remaining[%s] != keySerde.keySize[%s], buffer was the wrong size?!",
+            keyBuffer.remaining(),
+            keySize
+        );
+      }
+
+      final int prevRecordOffset = curWriteIndex * recordSize;
+      if (curWriteIndex == -1 || !keyEquals(keyBuffer, buffer, prevRecordOffset)) {
+        initNewSlot(keyBuffer);
+      }
+
+      final int curRecordOffset = curWriteIndex * recordSize;
+      for (int i = 0; i < aggregatorOffsets.length; i++) {
+        aggregators[i].aggregate(buffer, curRecordOffset + aggregatorOffsets[i]);
+      }
+
+      return AggregateResult.ok();
+    }
+    catch (Throwable t) {
+      finish();
+      throw t;
+    }
+  }
+
+  private boolean keyEquals(ByteBuffer curKeyBuffer, ByteBuffer buffer, int bufferOffset)
+  {
+    int i = 0;
+    for (; i + Long.BYTES <= keySize; i += Long.BYTES) {
+      if (curKeyBuffer.getLong(i) != buffer.getLong(bufferOffset + i)) {
+        return false;
+      }
+    }
+
+    for (; i + Integer.BYTES <= keySize; i += Integer.BYTES) {
```

Comment:
At most one iteration of this loop could happen, suggested to just remove it

Reply:
Good point. Removed.

Reply:
I added this check again without loop because this is a hot method called per each row.

Reply:
I wonder how does it compare to setting temporarily position and limit in buffer and calling buffer.equals() then, that may resort to some intrinsics, especially on Java 9. See https://stackoverflow.com/a/3769920/648955

Reply:
What kind of intrinsics do you mean? I'm seeing the implementation of ByteBuffer.equals() in Java 8 and 9 is

```java
    public boolean equals(Object ob) {
        if (this == ob)
            return true;
        if (!(ob instanceof ByteBuffer))
            return false;
        ByteBuffer that = (ByteBuffer)ob;
        if (this.remaining() != that.remaining())
            return false;
        int p = this.position();
        for (int i = this.limit() - 1, j = that.limit() - 1; i >= p; i--, j--)
            if (!equals(this.get(i), that.get(j)))
                return false;
        return true;
    }

    private static boolean equals(byte x, byte y) {
        return x == y;
    }
```